src/gsumm.c

```diff
@@ -1162,3 +1162,92 @@ SEXP gprod(SEXP x, SEXP narmArg) {
+        if (lag) {                                                                                                \
+          const int o = ff[i]-1+(grpn-thisn);                                                                     \
+          for (int j=0; j<thisn; ++j) {                                                                           \
+          const int k = issorted ? (o+j) : oo[o+j]-1;                                                             \
+            const CTYPE val = cycle ? (nosubset ? xd[k] : (irows[k]==NA_INTEGER ? fill : xd[irows[k]-1])) : fill; \
+            ASSIGN;                                                                                               \
+          }                                                                                                       \
+        }                                                                                                         \
+        for (int j=jstart; j<jend; ++j) {                                                                         \
+          const int k = issorted ? j : oo[j]-1;                                                                   \
+          const CTYPE val = nosubset ? xd[k] : (irows[k]==NA_INTEGER ? fill : xd[irows[k]-1]);                    \
+          ASSIGN;                                                                                                 \
+        }                                                                                                         \
+        if (!lag) {                                                                                               \
+          const int o = ff[i]-1;                                                                                  \
+          for (int j=0; j<thisn; ++j) {                                                                           \
+            const int k = issorted ? (o+j) : oo[o+j]-1;                                                           \
+            const CTYPE val = cycle ? (nosubset ? xd[k] : (irows[k]==NA_INTEGER ? fill : xd[irows[k]-1])) : fill; \
+            ASSIGN;                                                                                               \
+          }                                                                                                       \
+        }                                                                                                         \
+      }                                                                                                           \
+    }
+    switch(TYPEOF(x)) {
+      case LGLSXP:  { int *ansd=LOGICAL(tmp);             SHIFT(int,     LOGICAL,   ansd[ansi++]=val); } break;
+      case INTSXP:  { int *ansd=INTEGER(tmp);             SHIFT(int,     INTEGER,   ansd[ansi++]=val); } break;
+      case REALSXP: { double *ansd=REAL(tmp);             SHIFT(double,  REAL,      ansd[ansi++]=val); } break;
```

Comment:
Was just about to merge (looks awesome!) and then just saw `integer64` missed here (and `test_bit64` added to tests).  Bug fix 45 starts `shift(xInt64, fill=0)` so now that `shift` is optimized I wonder if the `shift` problems of int64 would return. Maybe existing tests of shift int64 don't test it by group then.
Can't think of anything else.

Reply:
Interesting. The `coerceAs` is for `fill` but `fill` isn't currently tested by the new gforce test 2224.  A loop of differently typed `fill` needs adding to the test then, and I suspect a specific case here for integer64 will then be needed to pass that?  Can I leave that to you, and the C99 warning, hence WIP again.